File: Camp Progress/207. Course Schedule II.cpp
// Question Link: https://leetcode.com/problems/course-schedule-ii/

// DFS Solution 1
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        for (int i = 0; i < prerequisites.size(); i++) {
            int from = prerequisites[i][1];
            int to = prerequisites[i][0];
            graph[from].push_back(to);
        }
        
        vector<bool> seen(numCourses, false);
        
        int putIndex = numCourses - 1;
        vector<int> result(numCourses, -1);
        
        for (int ithNode = 0; ithNode < numCourses; ithNode++) {
            if (!seen[ithNode]) {
                vector<bool> seenCycle(numCourses, false);
                bool cycleFound = dfs(ithNode, seen, seenCycle, putIndex, result, graph);
                if (cycleFound) {
                    return {};
                }
            }
        }
        
        return result;
    }
    
    bool dfs(int cur, vector<bool> &seen, vector<bool> &seenCycle, int &putIndex, vector<int> &result, vector<vector<int>> &graph) {
        seen[cur] = true;
        seenCycle[cur] = true;
        bool cycleFound = false;
        for (int &neighbor : graph[cur]) {
            if (seen[neighbor] == false) {
                cycleFound = dfs(neighbor, seen, seenCycle, putIndex, result, graph);
                if (cycleFound) {
                    return true;
                }
            }
            else if (seenCycle[neighbor] == true) {
                return true;
            }
        }
        
        result[putIndex--] = cur;
        seenCycle[cur] = false;
        return false;
    }
};

// DFS Solution 2 Using Unorderd Set
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        for (int i = 0; i < prerequisites.size(); i++) {
            int from = prerequisites[i][1];
            int to = prerequisites[i][0];
            graph[from].push_back(to);
        }
        
        vector<bool> seen(numCourses, false);
        int putIndex = numCourses - 1;
        vector<int> result(numCourses, -1);
        
        for (int ithNode = 0; ithNode < numCourses; ithNode++) {
            if (!seen[ithNode]) {
                unordered_set<int> seenCycle;
                bool cycleFound = dfs(ithNode, seen, seenCycle, putIndex, result, graph);
                if (cycleFound) {
                    return {};
                }
            }
        }
        
        return result;
    }
    
    bool dfs(int cur, vector<bool> &seen, unordered_set<int> &seenCycle, int &putIndex, vector<int> &result, vector<vector<int>> &graph) {
        seen[cur] = true;
        seenCycle.insert(cur);
        bool cycleFound = false;
        for (int &neighbor : graph[cur]) {
            if (seen[neighbor] == false) {
                cycleFound = dfs(neighbor, seen, seenCycle, putIndex, result, graph);
                if (cycleFound) {
                    return true;
                }
            }
            else if (seenCycle.find(neighbor) != seenCycle.end()) {
                return true;
            }
        }
        
        result[putIndex--] = cur;
        seenCycle.erase(cur);
        return false;
    }
};


// BFS Solution Kahn's Algorith
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> incomingEdges(numCourses, 0);
        for (int i = 0; i < prerequisites.size(); i++) {
            int from = prerequisites[i][1];
            int to = prerequisites[i][0];
            graph[from].push_back(to);
            incomingEdges[to]++;
        }
        
        queue<int> todo;
        for (int i = 0; i < numCourses; i++) {
            if (incomingEdges[i] == 0) {
                todo.push(i);
            }
        }
        
        vector<int> result;
        int totalCoursesProcesses = 0;
        while (!todo.empty()) {
            int cur = todo.front();
            todo.pop();
            result.push_back(cur);
            totalCoursesProcesses++;
            for (int &neighbor : graph[cur]) {
                incomingEdges[neighbor]--;
                if (incomingEdges[neighbor] == 0) {
                    todo.push(neighbor);
                }
            }
        }
        
        if (totalCoursesProcesses != numCourses) {
            result = {};
        }
        
        return result;
    }
};

File: Python3 Solutions/207. Course Schedule II.py
# Question Link: https://leetcode.com/problems/course-schedule-ii

# DFS Solution
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for x in range(numCourses)]
        for dest, src in prerequisites:
            graph[src].append(dest)
        
        index = [numCourses - 1]
        result = [0] * numCourses
        seen = [False] * numCourses
        
        for cur in range(numCourses):
            cycle = [False] * numCourses
            if seen[cur] == False:
                hasCycle = self.dfs(cur, graph, seen, cycle, result, index)
                if hasCycle:
                    return []
                
        return result
    
    def dfs(self, cur, graph, seen, cycle, result, index) -> bool:
        seen[cur] = True
        cycle[cur] = True
        
        for neigh in graph[cur]:
            if seen[neigh] == False:
                hasCycle = self.dfs(neigh, graph, seen, cycle, result, index)
                if hasCycle:
                    return True
                
            elif cycle[neigh]:
                return True
            
        result[index[0]] = cur
        index[0] -= 1
        cycle[cur] = False
        return False


# BFS Solution Kahn's Algorithm
from collections import deque

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for x in range(numCourses)]
        indegrees = [0] * numCourses
        for dest, src in prerequisites:
            graph[src].append(dest)
            indegrees[dest] += 1
            
        todo = deque()
        for i in range(numCourses):
            if indegrees[i] == 0:
                todo.append(i)
                
        result = []
        totalCoursesProcessed = 0
        while todo:
            cur = todo.pop()
            result.append(cur)
            totalCoursesProcessed += 1
            for neigh in graph[cur]:
                indegrees[neigh] -= 1
                if indegrees[neigh] == 0:
                    todo.append(neigh)
        
        if totalCoursesProcessed != numCourses:
            result = []
        
        return result